File: sdk/typescript/examples/02_transactions.ts
/**
 * KCM TypeScript SDK — Transaction Example.
 *
 * Demonstrates: begin, commit, and rollback scenarios with transactions.
 */

import { Database, FactData, KcmError, ErrorCode } from '../src/index';

console.log("=== KCM TypeScript SDK — Transaction Example ===\n");

const db = new Database();

// Insert baseline facts
db.insert({
    subject: 1, predicate: 0, object: 2, confidence: 0.95,
    evidence: 0, timestamp: 0, context: 0, version: 1, priority: 0, owner: 0,
});
db.insert({
    subject: 2, predicate: 1, object: 3, confidence: 0.90,
    evidence: 1, timestamp: 0, context: 1, version: 1, priority: 0, owner: 1,
});
console.log(`Initial: ${db.activeFactCount()} active facts\n`);

// --- COMMITTED TRANSACTION ---
console.log("--- Committed Transaction ---");
const txn1 = db.beginTransaction();
txn1.recordInsert({
    subject: 3, predicate: 2, object: 4, confidence: 0.85,
    evidence: 2, timestamp: 0, context: 2, version: 1, priority: 0, owner: 2,
}, db.factCount());
console.log("  Inserted fact in transaction");
txn1.commit();
console.log(`  After commit: ${db.activeFactCount()} active facts`);
console.assert(db.activeFactCount() === 3);

// --- ROLLED BACK TRANSACTION ---
console.log("\n--- Rolled Back Transaction ---");
const txn2 = db.beginTransaction();
txn2.recordInsert({
    subject: 4, predicate: 3, object: 5, confidence: 0.80,
    evidence: 3, timestamp: 0, context: 2, version: 1, priority: 0, owner: 3,
}, db.factCount());
console.log("  Inserted fact in transaction");
txn2.rollback();
console.log(`  After rollback: ${db.activeFactCount()} active facts`);
console.assert(db.activeFactCount() === 3);

// --- TRANSACTION WITH UPDATE ---
console.log("\n--- Transaction with Update ---");
const txn3 = db.beginTransaction();
txn3.recordUpdate(0, {
    subject: 10, predicate: 0, object: 20, confidence: 0.99,
    evidence: 5, timestamp: 0, context: 3, version: 2, priority: 0, owner: 1,
});
console.log("  Updated fact in transaction");
txn3.commit();
console.log("  Transaction committed successfully");

// --- TRANSACTION WITH DELETE ---
console.log("\n--- Transaction with Delete ---");
const countBefore: number = db.activeFactCount();
const txn4 = db.beginTransaction();
txn4.recordDelete(2);
console.log("  Deleted fact in transaction");
txn4.commit();
console.log(`  After commit: ${db.activeFactCount()} active facts (was ${countBefore})`);
console.assert(db.activeFactCount() === countBefore - 1);

// --- MULTIPLE OPERATIONS ---
console.log("\n--- Multiple Operations in Transaction ---");
const txn5 = db.beginTransaction();
txn5.recordInsert({
    subject: 10, predicate: 0, object: 20, confidence: 0.99,
    evidence: 0, timestamp: 0, context: 0, version: 1, priority: 0, owner: 0,
}, db.factCount());
txn5.recordInsert({
    subject: 30, predicate: 1, object: 40, confidence: 0.88,
    evidence: 0, timestamp: 0, context: 0, version: 1, priority: 0, owner: 0,
}, db.factCount() + 1);
txn5.recordInsert({
    subject: 50, predicate: 2, object: 60, confidence: 0.77,
    evidence: 0, timestamp: 0, context: 0, version: 1, priority: 0, owner: 0,
}, db.factCount() + 2);
console.log("  3 pending operations");
txn5.commit();
console.log(`  After commit: ${db.activeFactCount()} active facts`);

db.close();
console.log("\n=== All transaction operations completed ===");

File: sdk/python/examples/02_transactions.py
#!/usr/bin/env python3
"""KCM Python SDK — Transaction Example.

Demonstrates: begin, commit, and rollback scenarios with transactions.
"""

import os
import sys

sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "src"))

from kcm import Database, KcmError, ErrorCode


def main() -> None:
    print("=== KCM Python SDK — Transaction Example ===\n")

    db = Database()

    # Insert baseline facts
    db.insert(subject=1, predicate=0, object=2, confidence=0.95)
    db.insert(subject=2, predicate=1, object=3, confidence=0.90)
    print(f"Initial: {db.active_fact_count()} active facts\n")

    # --- COMMITTED TRANSACTION ---
    print("--- Committed Transaction ---")
    with db.begin_transaction() as txn:
        print(f"  Began {txn}")
        db.insert(subject=3, predicate=2, object=4, confidence=0.85)
        print(f"  Inserted fact in transaction")
        txn.commit()
        print(f"  Committed {txn}")
    print(f"  After commit: {db.active_fact_count()} active facts")
    assert db.active_fact_count() == 3

    # --- ROLLED BACK TRANSACTION ---
    print("\n--- Rolled Back Transaction ---")
    with db.begin_transaction() as txn:
        print(f"  Began {txn}")
        db.insert(subject=4, predicate=3, object=5, confidence=0.80)
        print(f"  Inserted fact in transaction")
        txn.rollback()
        print(f"  Rolled back {txn}")
    print(f"  After rollback: {db.active_fact_count()} active facts")
    assert db.active_fact_count() == 3

    # --- AUTO-ROLLBACK ON EXCEPTION ---
    print("\n--- Auto-Rollback on Exception ---")
    count_before = db.active_fact_count()
    try:
        with db.begin_transaction() as txn:
            db.insert(subject=5, predicate=4, object=6, confidence=0.70)
            raise ValueError("simulated error")
    except ValueError:
        pass
    print(f"  After exception: {db.active_fact_count()} active facts")
    print(f"  Transaction auto-rolled back: {db.active_fact_count() == count_before}")

    # --- MULTIPLE OPERATIONS IN TRANSACTION ---
    print("\n--- Multiple Operations in Transaction ---")
    with db.begin_transaction() as txn:
        db.insert(subject=10, predicate=0, object=20, confidence=0.99)
        db.insert(subject=30, predicate=1, object=40, confidence=0.88)
        db.insert(subject=50, predicate=2, object=60, confidence=0.77)
        print(f"  3 pending operations")
        txn.commit()
    print(f"  After commit: {db.active_fact_count()} active facts")
    assert db.active_fact_count() == 6

    db.close()
    print("\n=== All transaction operations completed ===")


if __name__ == "__main__":
    main()
